### ISSUE TITLE
File API

### ISSUE THREAD

**Issue by joelwurtz** [concurrent-php/ext-async#80]
Does API around file processing / event would likely be considered as a feature of this library ?

Or maybe it's already the case by using a file resource with the StreamWatcher api ? (but not sure since async api for files is generally a specific use case)

**Comment by martinschroeder** [comment]
Async filesystem operations are something that i will add to the extension. I would like to start with read & write streams to allow for async disc IO. There are a lot of filesystem functions provided by `libuv`, i think almost all of them can be implemented in the future.

**Comment by martinschroeder** [comment]
I started working on async file IO, and i am taking a different route with the implementation. Starting with PHP 7.3 RC5 it is possible to exchange PHP's default file stream wrapper with a custom implementation. I created a stream wrapper based on libuv and so far it works just fine, it only needs some more test cases.

The advantage of this approach is that most file IO operations (`fopen()`, `fwrite()`, `fread()`, `file_get_contents()`, etc.) are automatically async. Development happens on the `filesystem` branch, Windows is supported and i will compile a DLL once RC5 is released and the branch is merged.
